File: integrations/mem0/python/src/purra_mem0/direct_providers.py
"""Mem0's native text/embedding protocol backed by the current PurrA operation."""
from collections.abc import Mapping, Sequence

from purra.contracts import AgentMessage
from ._journal import MemoryError
from .providers import _execution


def _reject(execution):
    execution.stop("memory_provider_contract")
    raise MemoryError(execution.error)


class DirectLlm:
    def generate_response(self, messages, tools=None, tool_choice="auto", **kwargs):
        execution = _execution()
        execution.check()
        if (tools not in (None, []) or tool_choice != "auto"
                or set(kwargs) - {"response_format"}
                or kwargs.get("response_format") not in (None, {"type": "json_object"})
                or not isinstance(messages, Sequence) or isinstance(messages, (str, bytes))
                or not messages):
            _reject(execution)
        converted = []
        for message in messages:
            if (not isinstance(message, Mapping) or set(message) != {"role", "content"}
                    or message["role"] not in ("system", "user", "assistant")
                    or not isinstance(message["content"], str)):
                _reject(execution)
            converted.append(AgentMessage(role=message["role"], content=message["content"]))
        return execution.invoke("llm", tuple(converted))


class DirectEmbedder:
    def embed(self, text, memory_action=None):
        return self.embed_batch([text], memory_action)[0]

    def embed_batch(self, texts, memory_action="add"):
        execution = _execution()
        execution.check()
        if (memory_action not in (None, "add", "search", "update")
                or not isinstance(texts, Sequence) or isinstance(texts, (str, bytes))
                or any(not isinstance(text, str) for text in texts)):
            _reject(execution)
        return execution.invoke("embedding", tuple(texts)) if texts else []

File: integrations/mem0/typescript/src/direct-providers.ts
/** Mem0's native protocol; every call uses the currently bound PurrA operation. */
import type { Message } from "purra";
import { currentExecution, type ProviderExecution } from "./providers.js";
import { MemoryError } from "./journal.js";

function execution(): ProviderExecution {
  const value = currentExecution();
  if (!value) throw new MemoryError("memory_provider_unbound");
  value.check();
  return value;
}
function reject(value: ProviderExecution): never {
  value.stop("memory_provider_contract");
  throw new MemoryError(value.error!);
}

export class DirectLlm {
  async generateResponse(messages: unknown, format?: unknown, tools?: unknown, ...extra: unknown[]): Promise<string> {
    const value = execution();
    if (extra.length || (tools !== undefined && (!Array.isArray(tools) || tools.length))
      || (format !== undefined && (format === null || typeof format !== "object"
        || Object.keys(format).length !== 1 || (format as { type?: unknown }).type !== "json_object"))
      || !Array.isArray(messages) || !messages.length) reject(value);
    const converted: Message[] = Array.from(messages, (message: unknown) => {
      if (!message || typeof message !== "object" || Object.keys(message).sort().join() !== "content,role") reject(value);
      const m = message as { role: string; content: unknown };
      if (!["system", "user", "assistant"].includes(m.role) || typeof m.content !== "string") reject(value);
      return { role: m.role as "system" | "user" | "assistant", content: m.content };
    });
    return value.invoke("llm", converted);
  }
  async generateChat(): Promise<never> { return reject(execution()); }
}

export class DirectEmbedder {
  async embed(text: unknown, action?: unknown): Promise<number[]> { return (await this.embedBatch([text], action))[0]!; }
  async embedBatch(texts: unknown, action?: unknown): Promise<number[][]> {
    const value = execution();
    if ((action !== undefined && !["add", "search", "update"].includes(action as string))
      || !Array.isArray(texts) || Array.from(texts).some(text => typeof text !== "string")) reject(value);
    return texts.length ? value.invoke("embedding", [...texts]) : [];
  }
}
